### PR TITLE
SYCL : Move to compile time oneMKL interface backend selection for NVIDIA backend

### PR DESCRIPTION
**Pull request by s-Nick**
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

This patch move oneMKL interface calls to gemm from real time to compile time for NVIDIA backend, bringing improvements specially in text generation.

Tested on A100:
*Current*
| model                          |       size |     params | backend    | ngl |    sm | mmap |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ----: | ---: | ------------: | -------------------: |
| llama 70B Q4_K - Small         |  37.57 GiB |    70.55 B | SYCL       |  99 |  none |    0 |         pp512 |        705.51 ± 2.23 |
| llama 70B Q4_K - Small         |  37.57 GiB |    70.55 B | SYCL       |  99 |  none |    0 |         tg128 |         14.28 ± 0.05 |
| llama 8B Q4_K - Medium         |   4.58 GiB |     8.03 B | SYCL       |  99 |  none |    0 |         pp512 |      5426.17 ± 29.64 |
| llama 8B Q4_K - Medium         |   4.58 GiB |     8.03 B | SYCL       |  99 |  none |    0 |         tg128 |         81.36 ± 1.29 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | SYCL       |  99 |  none |    0 |         pp512 |      5592.87 ± 89.22 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | SYCL       |  99 |  none |    0 |         tg128 |         72.96 ± 0.91 |

build: 0f77aae (20)

*With changes*
| model                          |       size |     params | backend    | ngl | threads |    sm | mmap |          test |                  t/s |
| ------------------------------ | --------- | --------- | ---------- | -- | ------ | ---- | --- | ------------ | ------------------- |
| llama 70B Q4_K - Small         |  37.57 GiB |    70.55 B | SYCL       |  99 |       8 |  none |    0 |         pp512 |        720.68 ± 1.62 |
| llama 70B Q4_K - Small         |  37.57 GiB |    70.55 B | SYCL       |  99 |       8 |  none |    0 |         tg128 |         18.52 ± 0.07 |
| llama 8B Q4_K - Medium         |   4.58 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         pp512 |      5489.17 ± 30.44 |
| llama 8B Q4_K - Medium         |   4.58 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         tg128 |         91.99 ± 0.05 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         pp512 |     5439.13 ± 216.28 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         tg128 |         89.34 ± 0.05 |

build: ffd0a998 (4222)

### CODE REVIEW

**Comment by s-Nick** [comment]
@Alcpz Could you check it out?

**Comment by Alcpz** [comment]
@Rbiessy Feel free to give it a look, since you have experience working with oneMKL interface

File: ggml/src/ggml-sycl/dpct/helper.hpp [hunk 1]
```diff
@@ -1690,8 +1690,12 @@ namespace dpct
             auto data_b = get_memory<const Tb>(b);
             auto data_c = get_memory<Tc>(c);
             oneapi::mkl::blas::column_major::gemm(
-                q, a_trans, b_trans, m, n, k, alpha_value, data_a, lda,
-                data_b, ldb, beta_value, data_c, ldc);
+#ifdef GGML_SYCL_NVIDIA
```
Comment (NeoZhangJianyu) [inline ggml/src/ggml-sycl/dpct/helper.hpp:1693]: The macro make the code is hard to understand.
I suggest:

```
#ifdef GGML_SYCL_NVIDIA
        oneapi::mkl::blas::column_major::gemm(
                oneapi::mkl::backend_selector<oneapi::mkl::backend::cublas>{ q },
                a_trans, b_trans, m, n, k, alpha_value, data_a, lda,
                data_b, ldb, beta_value, data_c, ldc);
        }
#else
        oneapi::mkl::blas::column_major::gemm(
                q, 
                a_trans, b_trans, m, n, k, alpha_value, data_a, lda,
                data_b, ldb, beta_value, data_c, ldc);
        }
#endif
```

**Review by NeoZhangJianyu — COMMENTED** [review]
Same comments for other update.

**Comment by NeoZhangJianyu** [comment]
@s-Nick 
I guess the same method could help for Intel GPU.
Is it possible to update for Intel GPU too? like `oneapi::mkl::backend::mklgpu`.

**Review by Rbiessy — APPROVED** [review]
The oneMKL Interface changes look good to me.

File: ggml/src/ggml-sycl/dpct/helper.hpp (hunk 1 above)
Comment (NeoZhangJianyu) [reply ggml/src/ggml-sycl/dpct/helper.hpp:1693]: Please remember, the SYCL backend is initiated to support Intel GPU. :)
Support more vendor GPUs is added later.
The default code path should be optimized for Intel GPU.

It's OK to set special queue for other vendor GPUs.

**Comment by NeoZhangJianyu** [comment]
> Thank you for your review @NeoZhangJianyu Currently Intel GPU implementation uses oneMKL closed source library directly and it doesn't have nor need a `backend_selector`, therefore these changes aren't required or useful

OK, I see!

File: ggml/src/ggml-sycl/dpct/helper.hpp (hunk 1 above)
Comment (s-Nick) [reply ggml/src/ggml-sycl/dpct/helper.hpp:1692]: Code update for readability in  [f6e6fc4](https://github.com/ggerganov/llama.cpp/pull/10584/commits/f6e6fc4d47507ec8ffb0b7107d6f6d548824074d)

**Review by Alcpz — APPROVED** [review]
changes lgtm. Let's wait for the remaining thread to be resolved before merging.

**Review by NeoZhangJianyu — APPROVED** [review]

**Pull request merged**